Implement the next/previous editor commands and keybindings

leavengood opened the pull request:
This wraps around like VSCode.

Fixes #778 and part of #1423.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/onivim/oni2?pullRequest=1447) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/onivim/oni2?pullRequest=1447) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/onivim/oni2?pullRequest=1447) it.</sub>

leavengood commented:
Hmmm, I just did some testing in VSCode and I think this works a bit different than I thought. If you have just one editor group, it moves between the tabs in that group (which is what I implemented), if you have multiple editor groups it moves between them when you reach each end (with full wrap around.) That is probably why the command has `workspace` as a prefix.

This is probably still a good first step, but I may be able to implement the full version too at a later date, if we want it.

glennsl commented:
You should also run `esy format` before pushing to make sure the CI is happy.

glennsl commented:
> This is probably still a good first step, but I may be able to implement the full version too at a later date, if we want it.

Yes, I agree on both counts!

leavengood commented on src/Model/EditorGroup.re:
```diff
@@ -82,6 +82,38 @@ let _getAdjacentEditor = (editor: int, reverseTabOrder: list(int)) => {
+          | -1 => count - 1
+          // If this is past the end, go to zero, otherwise this index is fine
+          | i => i >= count ? 0 : i
+          };
```
I had that thought myself while working outside in my garden after making this change. I'll fix it.

glennsl replied:
Oh, you must live in one of those nice places. My garden is full of snow! No outside thinking for me...

leavengood replied:
Hahaha, I am in south Florida. Our weather was very nice this weekend, in the mid 70s during the day.

leavengood commented on src/Model/Actions.re:
```diff
@@ -128,6 +128,8 @@ type t =
   | StatusBar(StatusBarModel.action)
   | ViewCloseEditor(int)
+  | ViewNextEditor
+  | ViewPreviousEditor
```
OK that makes sense. It will annoy me not having type-checking on the string, and copy and pasting it everywhere, but I can change it until you add `CommandTriggered`.

leavengood commented on src/Model/EditorGroup.re:
```diff
@@ -82,6 +82,38 @@ let _getAdjacentEditor = (editor: int, reverseTabOrder: list(int)) => {
+    | Some(activeEditorId) =>
+      switch (_getIndexOfElement(activeEditorId, tabs)) {
+      | (-1) => model
+      | idx =>
```
Is there a good reason not to just resolve the TODO and use `List.find_opt`?

leavengood replied:
Overall since this was my first submission and I am still learning ReasonML, I was trying to avoid too much refactoring to add to the noise of the diff.

glennsl replied:
Yeah, I'm definitely not expecting you to do that. But if you want to, I'd encourage you to go for it.

> Is there a good reason not to just resolve the TODO and use List.find_opt?

Hmm, not that I can see. Apparently I'm the one who added that comment as well, but beats me why I did that instead of just changing it right away. Perhaps I wanted to avoid adding noise to the diff too. But it ought to be done at some point, so now seems as good a time as any.

leavengood replied:
It turns out `List.find_opt` is not quite what is needed. Given an editor id, it just returns Some(id), or None, which is pretty useless here since we want the index. In fact I put it into use without thinking about it, with some hilarious and confusing results. It happens to pass type checking since both the editor ID and the index are `int`.

I think I will just change `_getIndexOfElement` to return `option(int)`. I'm not sure why that wasn't done originally...

glennsl replied:
Ah, yeah good call! Dealing with indexes isn't a very natural use for lists, which is why something like this doesn't exist already in the standard library. It might be that the overall model here is a bit off, but just converting this to `option(int)` is a good way to address the immediate concern I think. :+1:

leavengood replied:
I agree we may want to adjust the model data structure given this and the later need for setting the active editor by index. I am not yet sure what that model should be. But as you say moving to `option(int)` improves things for now.

leavengood commented on src/Model/EditorGroup.re:
```diff
@@ -82,6 +82,38 @@ let _getAdjacentEditor = (editor: int, reverseTabOrder: list(int)) => {
+
+// The diff amounts are inverted because the list is in reverse order
+let nextEditor = setActiveEditorByIndexDiff(-1);
+let previousEditor = setActiveEditorByIndexDiff(1);
```
My primary goal was to avoid two very similar functions, and as you say it probably does not make sense to have diffs of more than -1 or 1. Plus something like `Next or `Previous won't need to comment about the inversion. I will try that.

leavengood commented on src/Model/EditorGroup.re:
```diff
@@ -82,6 +82,38 @@ let _getAdjacentEditor = (editor: int, reverseTabOrder: list(int)) => {
+  let tabs = model.reverseTabOrder;
+  let count = List.length(tabs);
+
+  if (count <= 1) {
```
It is really the case of 0 and 1 where I don't want to do anything, so checking the count makes some sense, and this felt like the best way to do that. Maybe I don't need this optimization if I make the rest of the function better.

glennsl replied:
Oh, sorry I missed that. In that case it might be a good idea to use pattern matching. That way you can combine the 0 and 1 case:

```reason
switch (tabs) {
| [] | [_] => model
| tabs => ...
};
```

Or leaving it as-is is fine too, but there's at least some ideas for you to work with.

leavengood commented:
I believe I have addressed most of the comments. I am still not very happy with how awkward the `setActiveEditorTo` function is in EditorGroup, but it works. The need for wrap around makes other ways of iterating through the list less workable.

glennsl commented on src/Model/EditorGroup.re:
```diff
@@ -61,16 +61,23 @@ let getOrCreateEditorForBuffer = (state, bufferId) => {
+        switch (_getIndexOfElement(elem, tl)) {
+        | None => None
+        | Some(i) => Some(i + 1)
+        }
```
You can use `Option.map` for this:

```suggestion
        _getIndexOfElement(elem, tl) |> Option.map(succ)
```

leavengood replied:
I figured there had to be something for that like in Rust but when looking at the Reason docs I did not see much. I need to check the OCaml docs more I suppose. Good tip with using `succ` too.

glennsl commented on src/Model/EditorGroup.re:
```diff
@@ -82,37 +89,47 @@ let _getAdjacentEditor = (editor: int, reverseTabOrder: list(int)) => {
+            // If this is past the end, go to zero
+          } else {
+            newIndex;
+          };
```
There are some convenience functions in `Core.Utility.IndexEx` for this:

https://github.com/onivim/oni2/blob/7493356cc4bae87343f291c792c4c0fc088fd8b8/src/Core/Utility/IndexEx.re#L1-L27

leavengood replied:
Yep, that is exactly what I need. I'll give it a try when I get a chance.

leavengood commented:
By the way, I just want to say I appreciate all the time and guidance on the reviews. I know it can be a lot of work, as I am frequently on the other side.

glennsl commented:
Happy to do it! I enjoy nitpicking :upside_down_face:

bryphe commented:
Such a great addition! Thank you for all the help with this @leavengood and @glennsl 🎉 

Can't wait to start using this.

bryphe commented:
Looking at the CI - looks like an intermittent test failure, not related to the changes - kicking off a new run.

bryphe commented:
/azp run

leavengood commented:
Thanks @bryphe! This is a really great project, I am amazed how nice it is already and I am glad to help how I can.

If CI passes I think we can merge this as is, and I can do a bit more clean-up when I work on the setting editor by number commands. I also would like to add a test for EditorGroup then as well.

leavengood commented:
I think this CI failure is not from my code again ☹️ 

Is there any way to speed up the CI? Taking 30-60 minutes seems pretty long to me. I don't know if this is something people can contribute to fix, of course it is probably very complicated with esy and dune, Azure, multiple operating systems, etc.

leavengood commented:
/azp run

glennsl commented:
/azp run

glennsl commented:
> Is there any way to speed up the CI?

I think @bryphe is pretty on top of this, constantly tweaking the Ci process, but it seems tricky. I think caching was brought back a few days ago, but it's often failed for us for various reasons. And the Windows CI is particularly slow and finicky, and often just hangs.

So I think it'd be hard to contribute to this, especially without a good overview of what's been tried before and what might seem to work in the short term but has proven to not work over time.

leavengood commented:
As noted now that CI passed I think we can merge this as is and I can do some of the adjustments Glenn recommended in future work.

bryphe commented:
> I think @bryphe is pretty on top of this, constantly tweaking the Ci process, but it seems tricky. I think caching was brought back a few days ago, but it's often failed for us for various reasons.

Yeah, the CI is always a challenge. When the caching works - it really speeds up the builds! But it tends to be fragile - all kinds of things can break it (ie, recently Azure Pipelines changed the default paths, which broke our cache directories). Some previous battles with the build cache: <https://github.com/onivim/oni2/pulls?q=is%3Apr+cache+is%3Aclosed>

I just noticed we weren't restoring the cache builds for OSX - so I opened a PR for that: https://github.com/onivim/oni2/pull/1458

One thing we are missing is we don't have a cache strategy at all for our one environment that builds on Docker (the `CentOS` build) - it'd be nice to have that cached, because it's the bottleneck if the caches are working everywhere else.

It can be contributed to for sure - the files that control it are the [`azure-pipelines.yml`](https://github.com/onivim/oni2/blob/master/azure-pipelines.yml) file and the [`.ci`] folder (https://github.com/onivim/oni2/tree/master/.ci) but it can be a grind! Tends to be very time consuming to test and iterate on fixes & changes.

> As noted now that CI passed I think we can merge this as is and I can do some of the adjustments Glenn recommended in future work.

Sounds great! Merging now 🎉 

Thanks @leavengood for all the work on this! It's an excellent contribution.
